compiler-core/src/analyse.rs

```diff
@@ -157,24 +157,29 @@ pub fn infer_module<A>(
         let statement = insert_type_alias(t, &mut env)?;
         typed_statements.push(statement);
     }
-    for c in statements.constants {
-        let statement = infer_module_constant(c, &mut env, &name)?;
-        typed_statements.push(statement);
-    }
 
-    // Sort the functions into dependency order for inference. Functions that do
-    // not depend on other functions are inferred first, then ones that depend
+    // Sort functions and constants into dependency order for inference. Definitions that do
+    // not depend on other definitions are inferred first, then ones that depend
     // on those, etc.
-    let function_groups = into_dependency_order(statements.functions)?;
+    let definition_groups = into_dependency_order(statements.functions, statements.constants)?;
     let mut working_group = vec![];
 
-    for group in function_groups {
+    for group in definition_groups {
         // A group may have multiple functions that depend on each other through
         // mutual recursion.
 
-        for function in group {
-            let inferred = infer_function(function, &mut env, &mut hydrators, &name)?;
-            working_group.push(inferred);
+        for definition in group {
+            match definition {
+                Definition::Function(f) => {
+                    let statement = infer_function(f, &mut env, &mut hydrators, &name)?;
+                    working_group.push(statement);
+                }
+                Definition::ModuleConstant(c) => {
+                    let statement = infer_module_constant(c, &mut env, &name)?;
+                    working_group.push(statement);
+                }
+                _ => unreachable!("Only funtions and constants should be in a definition group"),
```

Comment:
Rather than crash at runtime let's have some type that is either a function or a module constant, making the invalid state impossible.

Reply:
Got it. Will refactor this.